### ISSUE TITLE
TypeVar inference over union arguments is order-dependent and incorrectly collapses to one branch

### ISSUE THREAD

**Issue by AnJunHao** [astral-sh/ty#3139]
### Summary

```python
from typing import cast

class Parent[T]: ...

def func[T](param: Parent[T]) -> T:
    return cast(T, ...)

ca: Parent[str] = cast(Parent[str], Parent())
result_a = func(ca)
reveal_type(result_a) # Revealed type: `str` (revealed-type) [Ln 10, Col 13]

cb: Parent[int] = cast(Parent[int], Parent())
result_b = func(cb)
reveal_type(result_b) # Revealed type: `int` (revealed-type) [Ln 14, Col 13]

ca_or_cb: Parent[str] | Parent[int] = cast(Parent[str] | Parent[int], Parent())
result_c = func(ca_or_cb)
reveal_type(result_c) # Revealed type: `str` (revealed-type) [Ln 18, Col 13]

cb_or_ca: Parent[int] | Parent[str] = cast(Parent[int] | Parent[str], Parent())
result_d = func(cb_or_ca)
reveal_type(result_d) # Revealed type: `int` (revealed-type) [Ln 22, Col 13]

```

`func` correctly infers `T = str` for `Parent[str]` and `T = int` for `Parent[int]`, so when called with `Parent[str] | Parent[int]` the return type should be `str | int`, regardless of union member order. Instead, the checker reveals `str` for `Parent[str] | Parent[int]` and `int` for `Parent[int] | Parent[str]`, which suggests that type variable inference is selecting a single union arm rather than combining inference results across all arms. Since union order should not affect the inferred type, this appears to be a bug.

See playground:

[Playground](https://play.ty.dev/9aaba09b-0ce5-41d3-9573-9880bf9d7cf1)

### Version

ty 0.0.25

**Comment by carljm** [comment]
Thanks for the report!

So the most basic problem in this example is that `Parent` is inferred as bivariant in `T`, since `T` is unused. This means that `Parent[int]` and `Parent[str]` are equivalent types, which is why ty simplifies a union of those two types to an arbitrary choice of them. This is correct but confusing, which is why we plan (in #1728) to remove bivariance and just fall back to covariance in that case.

If you add a method returning `T` to the body of `Parent`, it becomes covariant in `T`. Then we no longer arbitrarily pick the first type in the union, but we then infer `Unknown` for `result_c` and `result_d`. I think this is basically #1703, a known limitation of the current solver.
